### PR TITLE
test: Arch/beng 470 validate eth genesis as needed for hive

### PR DESCRIPTION
**Pull request by archbear**

### CODE REVIEW

File: cosmos/runtime/baseapp/baseapp.go [hunk 1]
```diff
@@ -141,6 +145,28 @@ type PolarisBaseApp struct {
 	ERC20Keeper *erc20keeper.Keeper
 }
 
+// InitChainer initializes the chain.
+func (a *PolarisBaseApp) InitChainer(ctx sdk.Context, req abci.RequestInitChain) (abci.ResponseInitChain, error) {
+	// Validate that ethereum genesis and cosmos genesis timestamp are the same.
+	// TODO: investigate how we can make this work in evm/keeper/genesis.go
+	// currently doesn't work in there since the scope of gen is restricted to the evm component
+	if req.Time != ctx.BlockHeader().Time {
```
Comment (itsdevbear) [inline cosmos/runtime/baseapp/baseapp.go:154]: checking against the abci response isn't useful, you want to check against the ethereum genesis.

File: cosmos/runtime/baseapp/baseapp.go [hunk 2]
```diff
@@ -141,6 +148,33 @@ type PolarisBaseApp struct {
...
+	if err := json.Unmarshal(req.AppStateBytes, &genesisState); err != nil {
+		panic(err)
+	}
+
+	var ethGenesis *core.Genesis
+	if err := json.Unmarshal(genesisState["evm"], ethGenesis); err != nil {
+		return abci.ResponseInitChain{}, err
+	}
+
+	if ethGenesis.Timestamp != uint64(ctx.BlockHeader().Time.Unix()) {
```
Comment (itsdevbear) [inline cosmos/runtime/baseapp/baseapp.go:166]: You already have the genesis file, so tbh I'd just do this without using the context to reduce uncertainty

File: cosmos/x/evm/keeper/genesis.go [hunk 3]
```diff
@@ -58,3 +77,23 @@ func (k *Keeper) ExportGenesis(ctx sdk.Context) *types.GenesisState {
 	}
 	return genesisState
 }
+
+// validateEthGenesis is a InitGenesis helper which asserts that the ethereum genesis state is
+// consistent with the cosmos genesis state.
+func (k *Keeper) validateEthGenesis(ctx sdk.Context, genesisState types.GenesisState) error {
+	ethGenesis := enclib.MustUnmarshalJSON[core.Genesis]([]byte(genesisState.Params.EthGenesis))
+
+	if (ethGenesis.Coinbase != common.Address{0}) {
```
Comment (calbera) [inline cosmos/x/evm/keeper/genesis.go:86]:
```suggestion
	if (ethGenesis.Coinbase != common.Address{}) {
```

File: cosmos/x/evm/keeper/genesis.go [hunk 4]
```diff
@@ -58,3 +77,23 @@ func (k *Keeper) ExportGenesis(ctx sdk.Context) *types.GenesisState {
...
+		return fmt.Errorf("coinbase of the genesis block must be the null address, not: %s", ethGenesis.Coinbase.Hex())
+	}
+
+	// verify balances match
+	for addr, account := range ethGenesis.Alloc {
+		// no need to check for missing denom, since a missing denom will return 0 balance
+		coin := k.bk.GetBalance(ctx, lib.AddressToAccAddress(addr), genesisState.Params.EvmDenom)
+		if coin.Amount != math.NewIntFromBigInt(account.Balance) {
```
Comment (calbera) [inline cosmos/x/evm/keeper/genesis.go:94]: use the math.Int compare function rather than go equality

File: eth/core/genesis.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+// 	var p []struct{ Addr, Balance *big.Int }
+// 	if err := rlp.NewStream(strings.NewReader(data), 0).Decode(&p); err != nil {
+// 		panic(err)
+// 	}
+// 	ga := make(GenesisAlloc, len(p))
+// 	for _, account := range p {
+// 		ga[common.BigToAddress(account.Addr)] = GenesisAccount{Balance: account.Balance}
+// 	}
+// 	return ga
+// }
```
Comment (calbera) [inline eth/core/genesis.go:54]: is this code still needed?

**Review by calbera — APPROVED** [review]
utACK

**Comment by archbear** [comment]
image

**Comment by calbera** [comment]
my man

**Pull request closed without merging**